Ntseng/issue6

ntseng opened the pull request:
Summary
-------
- fixed a crash in changeclubmeeting
- changed club repeat period enums to the full word ("days" or "weeks")
- add Start Event button to reminder message
- cleaned up reminder and event scheduling logic
- added permissions check to `/club-config`

Local Tests Performed
---------------------
- open changeclubmeeting while nextMeeting is null
- used changeclubmeeting modal (will need migration for repeat period enums)
- started bot with next meetings set
- changed next meeting value via modal
- used /club-send-reminder

Issue
-----
Closes #6

ntseng committed:
#6: fix crash in changeclubmeeting

ntseng committed:
#6: change club repeat period enums to full word

ntseng committed:
#6: fix event start boundary calculations

ntseng committed:
#6: add startevent button to club reminders

ntseng committed:
#6: chain "event already finished" error correctly

ntseng committed:
#6: tighten reminder and event scheduling

ntseng committed:
#6: correct undefined type for eventId

SillySalamndr commented on source/modalSubmissions/changeclubmeeting.js:
```diff
@@ -16,17 +18,30 @@ module.exports = new ModalSubmission(id,
-				if (club.timeslot.periodCount && club.isRecruiting()) {
-					scheduleClubEvent(club, interaction.guild);
-				}
+			if (!nextMeetingInput) {
```
The vast majority of the start of this function is a _lot_ of input validation and parsing, which could be cleaned up with some early-out-like behavior. Additionally, it seems like if some of the input fails validation, we still call other helper functions to execute on the action before we display the errors to the user and leave, which seems wrong.

ntseng replied:
The only validations we can use before submission are based on string length or existence, so our checks for ability to parse to integer and ranges have to be in-app.

In the case where the modal receives some appropriate data and some errors that can't be caught prior to submission, I think it's best to still process the appropriate data.

SillySalamndr commented on source/helpers.js:
```diff
@@ -767,7 +753,6 @@ exports.setClubReminder = async function (club, channelManager) {
 
 /** The number of ms until a day before the next meeting, but not more than the max allowable setTimeout duration
  * @param {number} timestamp The unix timestamp of the next meeting (in seconds)
- * @returns
```
Was removing this line in the JSDocs intentional?

ntseng replied:
Yeah, type annotations were able to detect that the function was returning a number without it.

SillySalamndr commented on source/buttons/startevent.js:
```diff
@@ -0,0 +1,50 @@
+				interaction.message.edit({ components: [] });
+			});
+		} else {
+			interaction.reply({ content: "Only the club's leader or a moderator can start the event.", ephemeral: true });
```
Moving this validation error to the beginning as an early out would help tame the pretty intense amount of indentation in this function.

ntseng replied:
Added in https://github.com/Imaginary-Horizons-Productions/HorizonsBot/pull/44/commits/d87e3b6be573ee6843ceeed898d66572bcc0466f

ntseng committed:
#6: early out for startevent

ntseng committed:
#6: remove unused import

ntseng committed:
#6: add permissions check to /club-config

ntseng committed:
#6: missing import in /club-config

ntseng committed:
#6: add /club-send-reminder
Instead of automatically sending reminders for club meetings less than a day away on startup or configuration, let club leaders send manually.

ntseng committed:
#6: add no next meeting error to club send reminder
- also fix modifying event scheduling instead of reminder scheduling
- update default reminder to work with command

ntseng committed:
#6: tweak reminder message

ntseng committed:
Merge branch 'main' into ntseng/issue6
